[Unified Text Replacement] Undo-ing a replacement session can sometimes result in incorrect text

rr-codes opened the pull request:
#### 677afac9fefb30712341b5b909db687479ec7f53
<pre>
[Unified Text Replacement] Undo-ing a replacement session can sometimes result in incorrect text
<a href="https://bugs.webkit.org/show_bug.cgi?id=273331">https://bugs.webkit.org/show_bug.cgi?id=273331</a>
<a href="https://rdar.apple.com/127123467">rdar://127123467</a>

Reviewed by Aditya Keerthi.

When undo-ing a text replacement session, the original text may not be restored correctly. This is
because in `didEndTextReplacementSession`, the markers were being iterated from the beginning and
then their ranges were replaced with the original text. However, if the original text was a different
length than the marker range length, the replacement of the subsequent markers will end up using an
out-of-date replacement range.

Fix by iterating the markers backwards, since the replacement range of prior markers will never be
affected in this case.

Also perform a minor refactor to share more common logic.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::forEach&lt;DocumentMarkerController::IterationMode::Forward&gt;):
(WebCore::DocumentMarkerController::forEach&lt;DocumentMarkerController::IterationMode::Reversed&gt;):
(WebCore::DocumentMarkerController::forEachOfTypes):
(WebCore::DocumentMarkerController::forEach): Deleted.
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionPerformEditActionForPlainText):
(WebKit::UnifiedTextReplacementController::updateStateForSelectedReplacementIfNeeded):
(WebKit::UnifiedTextReplacementController::findReplacementMarkerByUUID):
(WebKit::UnifiedTextReplacementController::findReplacementMarkerContainingRange):
(WebKit::findReplacementMarkerByUUID): Deleted.
(WebKit::findReplacementMarkerContainingRange): Deleted.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:

Canonical link: <a href="https://commits.webkit.org/278069@main">https://commits.webkit.org/278069@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/405296ceaa07f46884dbebd52d90dcd3e98e63be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49353 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52386 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/25 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51657 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26255 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/52573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51453 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26182 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42508 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23637 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43691 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7721 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45549 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44195 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54104 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24434 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20620 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47643 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/capture-with-visibility-hidden-child.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25706 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42715 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46636 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/10850 "The change is no longer eligible for processing.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26545 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7096 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/dom/DocumentMarkerController.h:
```diff
@@ -51,6 +51,10 @@ class DocumentMarkerController final : public CanMakeCheckedPtr<DocumentMarkerCo
     WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR(DocumentMarkerController);
 public:
+    enum class IterationMode : bool {
+        Forward, Reversed
```
Nit - `Forward(s)` and `Backward(s)`, for consistency with other WebKit code.

Maybe `IterationDirection` too, for clarity.

whsieh commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -510,7 +511,30 @@ void DocumentMarkerController::forEach(const SimpleRange& range, OptionSet<Docum
+        auto markers = markersFor(node, types);
+
+        for (auto markerIterator = markers.rbegin(); markerIterator != markers.rend(); ++markerIterator) {
+            auto marker = *markerIterator;
```
What guarantees that the `marker` is non-null here?

aprotyas commented on Source/WebCore/dom/DocumentMarkerController.h:
```diff
@@ -93,7 +97,15 @@ class DocumentMarkerController final : public CanMakeCheckedPtr<DocumentMarkerCo
+    WEBCORE_EXPORT void forEach<IterationMode::Forward>(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&);
+
+    template<>
+    WEBCORE_EXPORT void forEach<IterationMode::Reversed>(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&);
```
Not that it can't be, but does it need to be templated? You could accept `IterationMode` as a function argument and then call separate implementations. Or if you want to avoid the runtime indirection:

```cpp
namespace IterationDirection {
    struct ForwardType {};
    struct BackwardType {};

    ForwardType Forward;
    BackwardType Backward;
};

void forEach(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&, IterationDirection::ForwardType);

void forEach(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&, IterationDirection::BackwardType);
```

And then you'd change callers as such:

```cpp
markers.forEach(..., WebCore::DocumentMarkerController::IterationDirection::Forward);

markers.forEach(..., WebCore::DocumentMarkerController::IterationDirection::Backward);
```

aprotyas replied:
This is basically the tag dispatch pattern, by the way. https://www.fluentcpp.com/2018/04/27/tag-dispatching/

rr-codes replied:
I think I'll stick with the template approach

rr-codes committed:
[Unified Text Replacement] Undo-ing a replacement session can sometimes result in incorrect text
https://bugs.webkit.org/show_bug.cgi?id=273331
rdar://127123467

Reviewed by Aditya Keerthi.

When undo-ing a text replacement session, the original text may not be restored correctly. This is
because in `didEndTextReplacementSession`, the markers were being iterated from the beginning and
then their ranges were replaced with the original text. However, if the original text was a different
length than the marker range length, the replacement of the subsequent markers will end up using an
out-of-date replacement range.

Fix by iterating the markers backwards, since the replacement range of prior markers will never be
affected in this case.

Also perform a minor refactor to share more common logic.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::forEach<DocumentMarkerController::IterationMode::Forward>):
(WebCore::DocumentMarkerController::forEach<DocumentMarkerController::IterationMode::Reversed>):
(WebCore::DocumentMarkerController::forEachOfTypes):
(WebCore::DocumentMarkerController::forEach): Deleted.
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionPerformEditActionForPlainText):
(WebKit::UnifiedTextReplacementController::updateStateForSelectedReplacementIfNeeded):
(WebKit::UnifiedTextReplacementController::findReplacementMarkerByUUID):
(WebKit::UnifiedTextReplacementController::findReplacementMarkerContainingRange):
(WebKit::findReplacementMarkerByUUID): Deleted.
(WebKit::findReplacementMarkerContainingRange): Deleted.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:

Canonical link: https://commits.webkit.org/278069@main
